File: python/9_finite-difference.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
QTT representation of finite-difference operator
"""

import matplotlib.pyplot as plt
import numpy as np


def main():
    depth = 4
    npoints = 2**depth
    cutoff = 1e-10
    max_rank = 4

    """target function and its derivative"""
    x = np.linspace(0.0, 1.0, npoints)
    y = np.exp(x)
    dy = np.exp(x)
    # y = np.cos(x)
    # dy = -np.sin(x)
    # y = np.exp(-(((x - 0.5) / 0.1) ** 2)) / (0.1 * np.sqrt(np.pi))
    # dy = -2.0 * (x - 0.5) * y / (0.1**2)

    """QTT decomposition of target function"""
    yt = y.copy()
    qtt = []
    rank = 1
    for k in range(0, depth - 1):
        yt = yt.reshape((rank * 2, -1))
        U, S, Vt = np.linalg.svd(yt, full_matrices=False)
        S = S[S > cutoff * S[0]]
        S = S[: min(len(S), max_rank)]
        rank_new = len(S)
        U = U[:, :rank_new]
        Vt = Vt[:rank_new, :]
        U = U.reshape((rank, 2, rank_new))
        qtt.append(U)
        yt = np.dot(np.diag(S), Vt)
        rank = rank_new
    yt = yt.reshape((rank, 2, 1))
    qtt.append(yt)
    print(f"QTT virtual dimensions: {[qtt[k].shape for k in range(depth)]}\n")

    """QTT for index shift operator"""
    s_qtt = []
    s = np.zeros((1, 2, 2, 2))
    s[0, 0, 0, 0] = s[0, 1, 1, 0] = s[0, 1, 0, 1] = s[0, 0, 1, 1] = 1.0
    s_qtt.append(s)
    for k in range(1, depth - 1):
        s = np.zeros((2, 2, 2, 2))
        s[0, 0, 0, 0] = s[0, 1, 1, 0] = s[0, 1, 0, 1] = s[1, 0, 1, 1] = 1.0
        s_qtt.append(s)
    s = np.zeros((2, 2, 2, 1))
    s[0, 1, 0, 0] = s[1, 0, 1, 0] = 1.0
    s_qtt.append(s)
    print(
        f"index shift operator virtual dimensions: {[s_qtt[k].shape for k in range(depth)]}\n"
    )

    """check for index shift operator"""
    if depth < 5:
        s_op = s_qtt[0]
        s_op = s_op.reshape(2, 2, 2)
        for k in range(1, depth):
            s_op = np.einsum("ijk,klmn->iljmn", s_op, s_qtt[k])
            s_op = s_op.reshape(
                s_op.shape[0] * s_op.shape[1],
                s_op.shape[2] * s_op.shape[3],
                s_op.shape[4],
            )
        s_op = s_op.reshape(s_op.shape[0], s_op.shape[1])
        print(f"index shift operator:\n{s_op}\n")

    """QTT for finite-difference operator"""
    d_qtt = []
    s = s_qtt[0].copy()
    st = s.transpose((0, 2, 1, 3))
    d = np.zeros((s.shape[0], 2, 2, 2 * s.shape[3]))
    d[:, :, :, : s.shape[3]] = s
    d[:, :, :, s.shape[3] : 2 * s.shape[3]] = -st  # minus sign for derivative
    d_qtt.append(d)
    for k in range(1, depth - 1):
        s = s_qtt[k].copy()
        st = s.transpose((0, 2, 1, 3))
        d = np.zeros((2 * s.shape[0], 2, 2, 2 * s.shape[3]))
        d[: s.shape[0], :, :, : s.shape[3]] = 2 * s
        d[s.shape[0] : 2 * s.shape[0], :, :, s.shape[3] : 2 * s.shape[3]] = 2 * st
        d_qtt.append(d)
    s = s_qtt[depth - 1].copy()
    st = s.transpose((0, 2, 1, 3))
    d = np.zeros((2 * s.shape[0], 2, 2, s.shape[3]))
    d[: s.shape[0], :, :, :] = 2 * s
    d[s.shape[0] : 2 * s.shape[0], :, :, :] = 2 * st
    d_qtt.append(d)
    print(
        f"finite-difference operator virtual dimensions: {[d_qtt[k].shape for k in range(depth)]}\n"
    )

    """check for finite-difference operator"""
    if depth < 5:
        d_op = d_qtt[0]
        d_op = d_op.reshape(2, 2, 4)
        for k in range(1, depth):
            d_op = np.einsum("ijk,klmn->iljmn", d_op, d_qtt[k])
            d_op = d_op.reshape(
                d_op.shape[0] * d_op.shape[1],
                d_op.shape[2] * d_op.shape[3],
                d_op.shape[4],
            )
        d_op = d_op.reshape(d_op.shape[0], d_op.shape[1])
        print(f"finite-difference operator:\n{d_op}\n")

    """apply finite-difference operator to target function"""
    dy_qtt = []
    for k in range(depth):
        t = np.einsum("ijkl,mjn->miknl", d_qtt[k], qtt[k])
        t = t.reshape(t.shape[0] * t.shape[1], t.shape[2], t.shape[3] * t.shape[4])
        dy_qtt.append(t)
    print(
        f"derivative QTT virtual dimensions: {[dy_qtt[k].shape for k in range(depth)]}\n"
    )

    """reconstruction"""
    dyr = dy_qtt[0]
    dyr = dyr.reshape(dyr.shape[0] * dyr.shape[1], dyr.shape[2])
    for k in range(1, depth):
        dyr = np.einsum("ij,jkl->ikl", dyr, dy_qtt[k])
        dyr = dyr.reshape(dyr.shape[0] * dyr.shape[1], dyr.shape[2])
    dyr = dyr.reshape(-1)

    """drop boundary points as derivative is not correct there due to boundary conditions"""
    x = x[1:-2]
    dy = dy[1:-2]
    dyr = dyr[1:-2]
    print(f"target derivative y' = {dy}\n")
    print(f"QTT derivative y' = {dyr}\n")
    print(f"error = {np.linalg.norm(dy - dyr) / len(dy)}\n")

    plt.figure()
    plt.plot(x, dy, label="target")
    plt.plot(x, dyr, label="QTT")
    plt.legend()
    plt.show()

    plt.figure()
    plt.plot(x, dy - dyr, label="error")
    plt.title("error")
    plt.legend()
    plt.show()


if __name__ == "__main__":
    main()

File: julia/9_finite-difference.jl
#!/usr/bin/env julia
"""
QTT representation of finite-difference operator (Julia)
"""

using LinearAlgebra
using TensorOperations
using Plots

function main()
    depth = 4
    npoints = 2^depth
    cutoff = 1e-10
    max_rank = 4

    # target function and its derivative
    x = range(0.0, 1.0; length=npoints) |> collect
    y = exp.(x)
    dy = exp.(x)
    # y = cos.(x); dy = -sin.(x)
    # y = @. exp(-(((x - 0.5)/0.1)^2)) / (0.1 * sqrt(pi)); dy = @. -2.0*(x-0.5)*y/(0.1^2)

    # QTT decomposition of target function
    yt = copy(y)
    qtt = Array{Float64}[]
    rank = 1
    for k in 0:(depth-2)
        yt = reshape(permutedims(reshape(yt, rank, :, 2), (3, 1, 2)), 2 * rank, :)
        F = svd(yt; full=false)
        S = F.S
        rank_new = findall(>(cutoff * S[1]), S)
        rank_new = min(length(rank_new), max_rank)
        U = F.U[:, 1:rank_new]
        Vt = F.Vt[1:rank_new, :]
        if k > 0
            U = permutedims(reshape(U, 2, rank, rank_new), (2, 1, 3))
        end
        push!(qtt, U)
        yt = Diagonal(S[1:rank_new]) * Vt
        rank = rank_new
    end
    push!(qtt, yt)
    println("QTT virtual dimensions: ", [size(qtt[k]) for k in 1:depth], "\n")

    # QTT for index shift operator
    s_qtt = Array{Float64}[]
    s = zeros(2, 2, 2)
    s[1, 1, 1] = 1.0
    s[2, 2, 1] = 1.0
    s[2, 1, 2] = 1.0
    s[1, 2, 2] = 1.0
    push!(s_qtt, s)
    for _ in 2:(depth-1)
        s = zeros(2, 2, 2, 2)
        s[1, 1, 1, 1] = 1.0
        s[1, 2, 2, 1] = 1.0
        s[1, 2, 1, 2] = 1.0
        s[2, 1, 2, 2] = 1.0
        push!(s_qtt, s)
    end
    s = zeros(2, 2, 2)
    s[1, 2, 1] = 1.0
    s[2, 1, 2] = 1.0
    push!(s_qtt, s)
    println("index shift operator virtual dimensions: ", [size(s_qtt[k]) for k in 1:depth], "\n")

    # check for index shift operator
    if depth < 5
        sop = s_qtt[1]
        for k in 2:(depth-1)
            # not [i,l,j,m,n] but [l,i,m,j,n] because of column-major order
            @tensor tmp[l, i, m, j, n] := sop[i, j, k_] * s_qtt[k][k_, l, m, n]
            sop = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 4) * size(tmp, 3), size(tmp, 5))
        end
        @tensor tmp[l, i, m, j] := sop[i, j, k] * s_qtt[depth][k, l, m]
        sop = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 4) * size(tmp, 3))
        println("index shift operator:\n", sop, "\n")
    end

    # QTT for finite-difference operator
    d_qtt = Array{Float64}[]
    s = copy(s_qtt[1])
    st = permutedims(s, (2, 1, 3))
    d = zeros(2, 2, 2 * size(s, 3))
    @views d[:, :, 1:size(s, 3)] .= -s # minus sign for derivative
    @views d[:, :, size(s, 3)+1:end] .= st
    push!(d_qtt, d)
    for k in 2:(depth-1)
        s = copy(s_qtt[k])
        st = permutedims(s, (1, 3, 2, 4))
        d = zeros(2 * size(s, 1), 2, 2, 2 * size(s, 4))
        @views d[1:size(s, 1), :, :, 1:size(s, 4)] .= 2 .* s
        @views d[size(s, 1)+1:end, :, :, size(s, 4)+1:end] .= 2 .* st
        push!(d_qtt, d)
    end
    s = copy(s_qtt[end])
    st = permutedims(s, (1, 3, 2))
    d = zeros(2 * size(s, 1), 2, 2)
    @views d[1:size(s, 1), :, :] .= 2 .* s
    @views d[size(s, 1)+1:end, :, :] .= 2 .* st
    push!(d_qtt, d)
    println("finite-difference operator virtual dimensions: ",
        [size(d_qtt[k]) for k in 1:depth], "\n")

    # check for finite-difference operator
    if depth < 5
        dop = d_qtt[1]
        for k in 2:(depth-1)
            # not [i,l,j,m,n] but [l,i,m,j,n] because of column-major order
            @tensor tmp[l, i, m, j, n] := dop[i, j, k_] * d_qtt[k][k_, l, m, n]
            dop = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 4) * size(tmp, 3), size(tmp, 5))
        end
        @tensor tmp[l, i, m, j] := dop[i, j, k] * d_qtt[depth][k, l, m]
        dop = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 4) * size(tmp, 3))
        println("finite-difference operator:\n", dop, "\n")
    end

    # apply finite-difference operator to target function
    dy_qtt = Array{Float64}[]
    D = d_qtt[1]
    Y = qtt[1]
    @tensor tmp[j, l, n] := D[j, k, l] * Y[k, n]
    push!(dy_qtt, reshape(tmp, size(tmp, 1), size(tmp, 3) * size(tmp, 2)))
    for k in 2:(depth-1)
        D = d_qtt[k]
        Y = qtt[k]
        # not [m,i,k,n,l] but [i,m,k,l,n] because of column-major order
        @tensor tmp[i, m, j, l, n] := D[i, j, k, l] * Y[m, k, n]
        push!(dy_qtt, reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 3), size(tmp, 5) * size(tmp, 4)))
    end
    D = d_qtt[end]
    Y = qtt[end]
    @tensor tmp[i, m, j] := D[i, j, k] * Y[m, k]
    push!(dy_qtt, reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 3)))
    println("derivative QTT virtual dimensions: ", [size(dy_qtt[k]) for k in 1:depth], "\n")

    # reconstruction
    dyr = dy_qtt[1]
    for k in 2:(depth-1)
        # not [a,c,d] but [c,a,d] because of column-major order
        @tensor tmp[c, a, d] := dyr[a, b] * dy_qtt[k][b, c, d]
        dyr = reshape(tmp, size(tmp, 2) * size(tmp, 1), size(tmp, 3))
    end
    @tensor tmp[a, c] := dyr[a, b] * dy_qtt[end][b, c]
    dyr = vec(transpose(tmp))

    # drop boundary points as derivative is not correct there due to boundary conditions
    x2 = x[2:end-1]
    dy2 = dy[2:end-1]
    dyr2 = dyr[2:end-1]
    if depth < 5
        println("target derivative y' = ", dy2, "\n")
        println("QTT derivative   y' = ", dyr2, "\n")
    end
    err = norm(dy2 .- dyr2) / length(dy2)
    println("error = ", err, "\n")

    p1 = plot(x2, dy2, label="target")
    plot!(p1, x2, dyr2, label="QTT")
    display(p1)
    readline()

    p2 = plot(x2, dy2 .- dyr2, label="error", title="error")
    display(p2)
    readline()
end

main()
